### ISSUE TITLE
Question: findOne with lazy relationship not working

### ISSUE THREAD

**Issue by plunchete** [typeorm/typeorm#2635]
**Issue type:**

[X] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[X] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[X] `0.2.7` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

```TypeScript
import {Entity, PrimaryGeneratedColumn, Column} from "typeorm";

@Entity({ name: 'users'})
export class User {
    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    email: string;
}

```

```TypeScript
import {Entity, PrimaryGeneratedColumn, OneToOne, JoinColumn } from 'typeorm';
import {User} from './User';

@Entity({ name: 'user_auth'})
export class UserInfo {
    @PrimaryGeneratedColumn()
    id: number;

    @OneToOne(() => User)
    @JoinColumn({ name: 'user_id'})
    user: User;
}
```

```TypeScript
const info = await this.userInfoRepository.findOne({ user: Promise.resolve(user) });
```

The logged query is
```
SELECT "UserInfo"."id" AS "UserInfo_id", "UserInfo"."user_id" AS "UseInfo_user_id" FROM "user_info" "UserInfo" WHERE "UserInfo"."user_id" = $1 -- PARAMETERS: [null]
```

What am I missing?
<!--
    To answer those questions you need to put "x" inside the square brackets, for example:
    [x] `mysql`
    [ ] `postgres`

    Also, please format your code properly (by taking code blocks into ```ts .... ```)
!>

**Comment by dominicdesu** [comment]
Is the find one method supposed to work with a Promise? What happens if you pass the user directly?

**Comment by plunchete** [comment]
No idea if it is supposed to work with a Promise but that compiles while passing the user directly results on:

<img width="833" alt="screen shot 2018-08-08 at 9 16 31 am" src="https://user-images.githubusercontent.com/50509/43850469-df089fb6-9aec-11e8-9e62-c72ab0499110.png">

**Comment by dominicdesu** [comment]
Hi,
is it possible that you've created the userInfoRepository by passing a Promise<User> to it?

**Comment by victoriqueko** [comment]
```

**Comment by afurculita** [comment]
@plunchete do you still have any issues with the query you've mentioned?

**Comment by afurculita** [comment]
Try the following:

```ts
const info = await this.userInfoRepository.findOne({ where: { user: user.id } });
```

**Comment by plunchete** [comment]
That one works but it results in two selects, one for the where and one for the id. 

Even if that approach works, should the Promise one work or even better allow a `user` without a promise in the `findOne`?

**Comment by plunchete** [comment]
The two queries that I mentioned were not taking into account the one to getting the `User`. It seems like typeorm issues a query to select by user.id and then another one to get by id, is that expected?

I don't think I understand what is the benefit of `@RelationId`. Does it work like lazy but without it being a promise?

**Comment by afurculita** [comment]
It is expected to have 2 queries if `user` is an unresolved Promise. To resolve the `user` Promise, TypeORM will have to do a query to get the data for building the `user` object. Only after that TypeORM can know what is the ID of the user.

But to circumvent this, I was proposing to try to get only the ID of the user and use only that.

Could you please show us first how do you get the `user` object? Where is it coming from?

**Comment by plunchete** [comment]
Again, I'm not talking about querying `User` but `UserInfo`. I think I figured it out, it seems to be part of the `save` operation that I call later in the code.

The user object is coming from
```
const user = await this.userRepository.findOne({ email });
```

BTW not that I'm not interested in a workaround, moving the relationship to eager is what I'm doing for now, but it seems like if the filter requires a promise it should use the id from the user passed inside or allow to send a user, right?

**Comment by afurculita** [comment]
Yes, the filter requires a promise around the user object and it should use the id from it, but it doesn't. I don't know if this is a design choice or there are some technical limitations, but it seems that it doesn't work with a Promise for now. 

A valid solution to this will be the one I suggested:
```ts
const info = await this.userInfoRepository.findOne({ where: { user: user.id } });
```
This allows you to keep `user` relation as a lazy one, with the Promise.
